### PR TITLE
pin_operator_digest: update plugin to work in tekton

### PR DESCRIPTION
**Pull request by hjmodi**
Remove worker and orchestrator differences completely,
and unify the login in the run() method of the plugin.
In OSBS 2 there will not be any worker orchestrator
abstraction.

* CLOUDBLD-8112

Signed-off-by: Harsh Modi <hmodi@redhat.com>

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] JSON/YAML configuration changes are updated in the relevant schema
- [ ] Changes to metadata also update the documentation for the metadata
- [ ] Pull request has a link to an osbs-docs PR for user documentation updates
- [ ] New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/plugins/pre_pin_operator_digest.py [hunk 1]
```diff
@@ -67,8 +62,7 @@ def __init__(self, workflow, replacement_pullspecs=None,
         Initialize pin_operator_digests plugin
 
         :param workflow: DockerBuildWorkflow instance
-        :param replacement_pullspecs: Dict[str, str], computed in orchestrator,
-                                      provided to workers by osbs-client
+        :param replacement_pullspecs: Dict[str, str]
```
Comment (chmeliik) [inline atomic_reactor/plugins/pre_pin_operator_digest.py:65]: This param can be removed, and ideally the `self.replacement_pullspecs` attribute as well (now, `replacement_pullspecs` will be just be computed in the run method)

**Review by chmeliik — COMMENTED** [review]
Nice, now the only necessary changes should be to `_get_operator_manifest()` from a build_dir rather than the source dir and make the changes (pretty much just `operator_csv.dump()`?) in all build dirs

File: atomic_reactor/plugins/pre_pin_operator_digest.py [hunk 2]
```diff
@@ -267,35 +227,17 @@ def run_in_orchestrator(self):
         else:
             if pullspecs:
                 replacement_pullspecs = self._get_replacement_pullspecs(pullspecs)
```
Comment (rcerven) [inline atomic_reactor/plugins/pre_pin_operator_digest.py:229]: this variable should be set to {} or None before IF

File: atomic_reactor/plugins/pre_pin_operator_digest.py [hunk 3]
```diff
@@ -267,35 +227,17 @@ def run_in_orchestrator(self):
...
-
-        Exclude CSVs which already have a relatedImages section.
-        """
-        operator_manifest = self._get_operator_manifest()
-
-        if self._skip_all():
-            self.log.warning("skip_all defined, not running on worker")
-            return
-
         replacement_pullspecs = {
```
Comment (rcerven) [inline atomic_reactor/plugins/pre_pin_operator_digest.py:235]: you should use replacement_pullspecs which you've got previously

File: atomic_reactor/plugins/pre_pin_operator_digest.py [hunk 4]
```diff
@@ -318,6 +260,23 @@ def run_in_worker(self):
...
+
+        :return: bool, should plugin run?
+        """
+        if not has_operator_bundle_manifest(self.workflow):
+            self.log.info("Not an operator manifest bundle build, skipping plugin")
+            return False
+        if not self.workflow.conf.operator_manifests:
+            msg = "operator_manifests configuration missing in reactor config map, aborting"
+            self.log.warning(msg)
+            return False
```
Comment (rcerven) [inline atomic_reactor/plugins/pre_pin_operator_digest.py:277]: I wonder wonder shouldn't this actually fail? I mean you want to use feature so you expect it to work, but this way we are just silently doing nothing if it isn't configured

@chmeliik @MartinBasti

Comment (chmeliik) [reply atomic_reactor/plugins/pre_pin_operator_digest.py:277]: Hmm, hard to say. The current behavior is not ideal, but immediately failing if the configuration is missing may not be ideal either. Users have ways to do the pinning themselves and tell OSBS to skip processing, so maybe that should be evaluated before checking for server config and raising errors.

TL;DR, I don't think it should be fixed in this PR

File: atomic_reactor/plugins/pre_pin_operator_digest.py (hunk 1 above)
Comment (hjmodi) [reply atomic_reactor/plugins/pre_pin_operator_digest.py]: updated.

File: atomic_reactor/plugins/pre_pin_operator_digest.py (hunk 2 above)
Comment (hjmodi) [reply atomic_reactor/plugins/pre_pin_operator_digest.py:226]: updated.

File: atomic_reactor/plugins/pre_pin_operator_digest.py (hunk 3 above)
Comment (hjmodi) [reply atomic_reactor/plugins/pre_pin_operator_digest.py:232]: that one is a list and the latest parts require a dictionary so this step converts it.

File: atomic_reactor/plugins/pre_pin_operator_digest.py [hunk 5]
```diff
@@ -451,18 +410,11 @@ def _get_replacement_pullspecs_from_csv_modifications(self, pullspecs):
         pullspecs_set = set(pullspecs)
         mod_pullspecs_set = set((ImageName.parse(p['original']) for p in mod_pullspec_repl))
 
-        missing = pullspecs_set - mod_pullspecs_set
-        if missing:
+        diff = pullspecs_set - mod_pullspecs_set
```
[the remark this answers is not in the document]
Comment (hjmodi) [reply atomic_reactor/plugins/pre_pin_operator_digest.py]: updated the testcase to do the same.

**Comment by hjmodi** [comment]
> Nice, now the only necessary changes should be to `_get_operator_manifest()` from a build_dir rather than the source dir and make the changes (pretty much just `operator_csv.dump()`?) in all build dirs

hmm, let me look into it. My assumption was that we're getting config from the source repo which should ideally be in the build_dir only so we don't have to worry about it here.

I'll try to ensure that's the case.

**Comment by chmeliik** [comment]
> hmm, let me look into it. My assumption was that we're getting config from the source repo which should ideally be in the build_dir only so we don't have to worry about it here.
> 
> I'll try to ensure that's the case.

Config can be read from the source repo, but the manifest files themselves should probably be read from one of the platform-specific build dirs. Though it might be fine to *read* manifests from the source repo as long as all *write* operations are done only in build dirs, so if it's not an easy change to make it's probably fine

File: tests/plugins/test_pin_operator_digests.py [hunk 6]
```diff
@@ -713,26 +713,6 @@ def test_replacement_opt_out(self, pin_digest, replace_repo, replace_registry,
...
-        original_content = csv.read_text("utf-8")
-
-        user_config = get_user_config(OPERATOR_MANIFESTS_DIR)
-        runner = mock_env(workflow, source_dir, site_config=get_site_config(),
-                          user_config=user_config)
-        runner.run()
-
-        assert "Replacing pullspecs" not in caplog.text
-        assert "Creating relatedImages section" not in caplog.text
-        assert csv.read_text("utf-8") == original_content
```
Comment (chmeliik) [inline tests/plugins/test_pin_operator_digests.py:724]: This test checked that OSBS will skip CSV files if they already have a relatedImages section. Is that still tested somewhere?

**Review by chmeliik — COMMENTED** [review]
@hjmodi the multi-arch handling still needs to be added, right? The plugin will have to use `workflow.build_dir.for_each_platform()` or `for_all_platforms_copy()` to make changes in build dirs

File: tests/plugins/test_pin_operator_digests.py (hunk 6 above)
Comment (hjmodi) [reply tests/plugins/test_pin_operator_digests.py:724]: It's sort of a weird situation with this testcase. It was ensuring that the `run_in_worker` didn't fail if there was a conflict between `relateImages` and env variables of the same since it should have been caught in `run_in_orchestrator`(it throws run time error if skip_all is False and if it's True the worker doesn't run.). This didn't make sense to me since `run_in_orchestrator` was always running before `run_in_worker` and this testcase was simulating something that wasn't actually supposed to happen.

Furthermore, in OSBS2 `run_in_orchestrator` and `run_in_worker` are merged in the `run` method so `run_in_orchestrator` steps would always execute before `run_in_worker` steps which invalidate this testcase.

Comment (chmeliik) [reply tests/plugins/test_pin_operator_digests.py:724]: I think the comment at the top is the confusing part here. The test doesn't check the conflict handling, it checks that the worker part of the plugin will not touch CSV files if they already have `relatedImages`, regardless of whether there is a conflict or not. I think the fix for this test would be to remove the `has_envs` parametrization and set it to False (so that there is no conflict). The conflict handling should already have its own test I think.

Comment (hjmodi) [reply tests/plugins/test_pin_operator_digests.py:724]: Oh I see. It makes sense now. Updated the testcase instead of removing it.

**Comment by hjmodi** [comment]
> @hjmodi the multi-arch handling still needs to be added, right? The plugin will have to use `workflow.build_dir.for_each_platform()` or `for_all_platforms_copy()` to make changes in build dirs

updated the code to handle multi-arch :)

File: atomic_reactor/plugins/pre_pin_operator_digest.py [hunk 7]
```diff
@@ -341,9 +303,13 @@ def _skip_all(self):
...
-        manifests_dir = self.workflow.source.manifests_dir
-        self.log.info("Looking for operator CSV files in %s", manifests_dir)
-        operator_manifest = OperatorManifest.from_directory(manifests_dir)
+        if self.workflow.source.config.operator_manifests is None:
+            raise RuntimeError("operator_manifests configuration missing in container.yaml")
+        self.log.info("Looking for operator CSV files in %s", self.workflow.build_dir.path)
+        manifests_dir = self.workflow.source.config.operator_manifests["manifests_dir"]
+        kwargs = {'repo_dir': self.workflow.build_dir.any_build_dir.path}
+        operator_manifest = OperatorManifest.from_directory(
+            os.path.join(self.workflow.build_dir.any_build_dir.path, manifests_dir), **kwargs)
```
Comment (rcerven) [inline atomic_reactor/plugins/pre_pin_operator_digest.py:312]: why are you passing kwargs again with the same path?
originally there was just OperatorManifest.from_directory(manifests_dir)

**Review by rcerven — APPROVED** [review]

**Review by lkolacek — APPROVED** [review]

File: atomic_reactor/plugins/pre_pin_operator_digest.py [hunk 8]
```diff
@@ -302,9 +303,13 @@ def _skip_all(self):
                                "bundles".format(component))
 
     def _get_operator_manifest(self):
-        manifests_dir = self.workflow.source.manifests_dir
-        self.log.info("Looking for operator CSV files in %s", manifests_dir)
-        operator_manifest = OperatorManifest.from_directory(manifests_dir)
+        if self.workflow.source.config.operator_manifests is None:
+            raise RuntimeError("operator_manifests configuration missing in container.yaml")
+        self.log.info("Looking for operator CSV files in %s", self.workflow.build_dir.path)
+        manifests_dir = self.workflow.source.config.operator_manifests["manifests_dir"]
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply atomic_reactor/plugins/pre_pin_operator_digest.py:309]: Filed internal tracker
